src/kcas.ml

```diff
@@ -13,165 +13,163 @@ module Id = struct
-  let r = rdcss_read a.content in
-  match r with
-  | CASN_DESC c ->
-      ignore @@ casn_proceed c;
-      get a
-  | WORD out -> out
-  | _ -> assert false
-
-let kCAS ?(presort = true) cas_list =
-  match cas_list with
+let set atom value =
+  Atomic.set atom.state { before = value; after = value; casn = casn_after }
+
+let rec release_after = function
+  | NIL -> true
+  | CAS (_, state, lt, gt) ->
+      state.casn <- casn_after;
+      state.before <- state.after;
+      release_after lt |> ignore;
+      release_after gt
+
+let rec release_before = function
+  | NIL -> false
+  | CAS (_, state, lt, gt) ->
+      state.casn <- casn_before;
+      state.after <- state.before;
+      release_before lt |> ignore;
+      release_before gt
+
+(* Note: The writes to `cas.casn <- ...` above could be removed to reduce
+   time at the cost of increasing space usage (by a constant factor). *)
+
+let release cass = function
+  | `After -> release_after cass
+  | `Before -> release_before cass
+
+let finish casn (`Undetermined cass as undetermined)
+    (status : [ `Before | `After ]) =
+  if Atomic.compare_and_set casn (undetermined :> status) (status :> status)
+  then release cass status
+  else Atomic.get casn == `After
+
+let rec determine casn undetermined skip = function
+  | NIL -> skip || finish casn undetermined `After
+  | CAS (atom, state, lt, gt) as eq ->
+      determine casn undetermined true lt
+      &&
+      let state' = Atomic.get atom.state in
+      if state == state' then determine casn undetermined skip gt
+      else
+        let before = state.before in
+        let before' = state'.before and after' = state'.after in
+        if
+          (before == before' && before == after')
+          || before == if is_after state'.casn then after' else before'
+        then
+          let status = Atomic.get casn in
+          if status != (undetermined :> status) then status == `After
+          else if Atomic.compare_and_set atom.state state' state then
+            determine casn undetermined skip gt
+          else determine casn undetermined skip eq
+        else finish casn undetermined `Before
+
+and is_after casn =
+  match Atomic.get casn with
+  | `Undetermined cass as undetermined -> determine casn undetermined false cass
+  | `After -> true
+  | `Before -> false
+
+let cas atom before after =
+  let state = { before = after; after; casn = casn_after } in
+  let state' = Atomic.get atom.state in
+  let before' = state'.before and after' = state'.after in
+  ((before == before' && before == after')
+  || before == if is_after state'.casn then after' else before')
+  && Atomic.compare_and_set atom.state state' state
+
+let commit (T (r, expect, update)) = cas r expect update
+
+let get atom =
+  let state = Atomic.get atom.state in
+  let before = state.before and after = state.after in
+  if before == after || is_after state.casn then after else before
+
+let overlap () = failwith "kcas: location overlap" [@@inline never]
```

Comment:
The `overlap` function should never be called in a correct program.  Call of `overlap ()` rather than `failwith "..."` generates slightly shorter code on some platforms.  This potentially slightly reduces the cost of the `splay` function.  Probably doesn't make a big difference, but I have noticed that at least on Apple M1 code size/alignment may have large impact on performance.